trieste/acquisition/rule.py

```diff
@@ -1484,122 +1484,113 @@ def filter_datasets(
         return filtered_datasets
 
 
-class UpdatableTrustRegionBox(Box, UpdatableTrustRegion):
-    """
-    A simple updatable box search space with a centre location and an associated global search
-    space.
-    """
+class UpdatableTrustRegionWithLocationInGlobalSearchSpace(UpdatableTrustRegion):
+    """An updatable trust region with a concept of a location within a global search space."""
 
-    def __init__(
+    def _init_location(
         self,
-        global_search_space: SearchSpace,
-        region_index: Optional[int] = None,
-        input_active_dims: Optional[Union[slice, Sequence[int]]] = None,
-    ):
-        """
-        :param global_search_space: The global search space this search space lives in.
-        :param region_index: The index of the region in a multi-region search space. This is used to
-            identify the local models and datasets to use for acquisition. If `None`, the
-            global models and datasets are used.
-        :param input_active_dims: The active dimensions of the input space, either a slice or list
-            of indices into the columns of the space. If `None`, all dimensions are active.
-        """
-        Box.__init__(self, global_search_space.lower, global_search_space.upper)
-        UpdatableTrustRegion.__init__(self, region_index, input_active_dims)
-        self._global_search_space = global_search_space
-        # Random initial location in the global search space.
-        self.location = tf.squeeze(global_search_space.sample(1), axis=0)
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+        location_candidate: Optional[TensorType] = None,
+    ) -> None:
+        # Initialize the location of the region.
+        # If no candidate is provided, take a random location from the global search space.
+        if location_candidate is not None:
+            self.location = location_candidate
+        else:
+            self.location = tf.squeeze(self.global_search_space.sample(1), axis=0)
 
     @property
+    @abstractmethod
     def location(self) -> TensorType:
-        """The centre of the box."""
-        return self._location
+        """The centre of the region."""
 
     @location.setter
+    @abstractmethod
     def location(self, location: TensorType) -> None:
-        self._location = location
+        """Set the centre of the region."""
 
     @property
+    @abstractmethod
     def global_search_space(self) -> SearchSpace:
-        """The global search space this search space lives in."""
-        return self._global_search_space
+        """The global search space this region lives in."""
 
 
-class SingleObjectiveTrustRegionBox(UpdatableTrustRegionBox):
-    """An updatable box search space for use with trust region acquisition rules."""
+# TODO: what is a good name for this class?
```

Comment:
Can you explain the type hierarchy here? AFAICT there are the following (and probably some more I've missed):

1. `UpdatableTrustRegion(SearchSpace)`
    1. `UpdatableTrustRegionWithLocationInGlobalSearchSpace(UpdatableTrustRegion)` — are there any updatable trust regions that _don't_ have a location in global search space? or can we merge this with the parent class?
        1. `UpdatableTrustRegionBox(Box, UpdatableTrustRegionWithLocationInGlobalSearchSpace)`
        1. `UpdatableTrustRegionDiscrete(DiscreteSearchSpace, UpdatableTrustRegionWithLocationInGlobalSearchSpace)`
        1. `GenericUpdatableTrustRegion(UpdatableTrustRegionWithLocationInGlobalSearchSpace)` — is the point of this to add the mechanism for defining single objective updatable trust regions for use with trust region acquisition rules? If so, maybe call it `SingleObjectiveTrustRegion`?
            1. `SingleObjectiveTrustRegionBox(UpdatableTrustRegionBox, GenericUpdatableTrustRegion)`
            1. `SingleObjectiveTrustRegionDiscrete(UpdatableTrustRegionDiscrete, GenericUpdatableTrustRegion)`
    1. `UpdatableTrustRegionProduct(TaggedProductSearchSpace, UpdatableTrustRegion)`

Reply:
That is the correct hierarchy. The main concept is that `GenericUpdatableTrustRegion` implements the core algorithm, and depends on location/global-space but doesn't have an implementation for them. Classes `UpdatableTrustRegionBox` and `UpdatableTrustRegionDiscrete` provide this implementation of location/global-space for different types of regions. These ingredients are brought together in the final concrete classes `SingleObjectiveTrustRegion*`.

> ... can we merge this with the parent class?
The intention was for `UpdatableTrustRegion` to be more generic. We don't have any trust-regions without location etc., but @hstojic uses these classes for completely different use. So after discussion with @hstojic, the right answer probably is to rename `UpdatableTrustRegion` as `UpdatableSearchSpace` and move it to `space.py`; and then drop `WithLocationInGlobalSearchSpace` from the other class (plus some other relate minor refactor).

> ... If so, maybe call it SingleObjectiveTrustRegion?
The `GenericUpdatableTrustRegion` now support algorithms with multiple datasets, so not just single objective trust regions. So that is not a suitable name.

Reply:
Actually, I don't think I will move `UpdatableSearchSpace` to `space.py`. It depends on `Dataset` and `Tag` etc., and that would add extra dependencies to that file.